File: RPi/pyRF24Network/examples/helloworld_tx.py
"""Simplest possible example of using RF24Network in RX role.
Sends messages from to receiver.
"""
import time
import struct
from RF24 import RF24
from RF24Network import RF24Network, RF24NetworkHeader


########### USER CONFIGURATION ###########
# See https://github.com/TMRh20/RF24/blob/master/pyRF24/readme.md
# Radio CE Pin, CSN Pin, SPI Speed
# CE Pin uses GPIO number with BCM and SPIDEV drivers, other platforms use
# their own pin numbering
# CS Pin addresses the SPI bus number at /dev/spidev<a>.<b>
# ie: RF24 radio(<ce_pin>, <a>*10+<b>); spidev1.0 is 10, spidev1.1 is 11 etc..

# Generic:
radio = RF24(22, 0)
################## Linux (BBB,x86,etc) #########################
# See http://nRF24.github.io/RF24/pages.html for more information on usage
# See http://iotdk.intel.com/docs/master/mraa/ for more information on MRAA
# See https://www.kernel.org/doc/Documentation/spi/spidev for more
# information on SPIDEV

# instantiate the network node using `radio` object
network = RF24Network(radio)

# Address of our node in Octal format (01,021, etc)
this_node = 0o1

# Address of the other node
other_node = 0o0

# How long to wait before sending the next message
interval = 2000  # in milliseconds

# initialize the radio
if not radio.begin():
    raise RuntimeError("radio hardware not responding")

radio.channel = 90

# initialize the network node
network.begin(this_node)

# radio.printDetails()
radio.printPrettyDetails()
packets_sent = 0
last_sent = 0

try:
    while True:
        network.update()
        now = int(time.monotonic_ns() / 1000000)
        # If it's time to send a message, send it!
        if now - last_sent >= interval:
            last_sent = now
            packets_sent += 1
            payload = struct.pack("<LL", now, packets_sent)
            ok = network.write(RF24NetworkHeader(other_node), payload)
            print(f"Sending {packets_sent}...", "ok." if ok else "failed.")
except KeyboardInterrupt:
    print("powering down radio and exiting.")
    radio.powerDown()

File: examples_RPi/helloworld_tx.cpp
/*
 Update 2014 - TMRh20
 */

/**
 * Simplest possible example of using RF24Network,
 *
 * RECEIVER NODE
 * Listens for messages from the transmitter and prints them out.
 */

//#include <cstdlib>
#include <RF24/RF24.h>
#include <RF24Network/RF24Network.h>
#include <iostream>
#include <ctime>
#include <stdio.h>
#include <time.h>

using namespace std;

RF24 radio(22, 0); // (CE Pin, CSN Pin, [SPI Speed (in Hz)])

RF24Network network(radio);

// Address of our node in Octal format (01,021, etc)
const uint16_t this_node = 01;

// Address of the other node
const uint16_t other_node = 00;

// How often (in milliseconds) to send a message to the `other_node`
const unsigned long interval = 2000;

uint32_t last_sent;    // When did we last send?
uint32_t packets_sent; // How many have we sent already

struct payload_t
{ // Structure of our payload
    uint32_t ms;
    uint32_t counter;
};

int main(int argc, char** argv)
{
    // Refer to RF24 docs or nRF24L01 Datasheet for settings

    if (!radio.begin()) {
        printf("Radio hardware not responding!\n");
        return 0;
    }

    delay(5);
    radio.setChannel(90);
    network.begin(/*node address*/ this_node);
    radio.printDetails();

    while (1) {

        network.update();
        uint32_t now = millis(); // If it's time to send a message, send it!
        if (now - last_sent >= interval) {
            last_sent = now;

            printf("Sending ..\n");
            payload_t payload = {millis(), packets_sent++};
            RF24NetworkHeader header(/*to node*/ other_node);
            bool ok = network.write(header, &payload, sizeof(payload));
            printf("%s.\n", ok ? "ok" : "failed");
        }
    }

    return 0;
}
